atomic_reactor/plugins/prepub_flatpak_create_oci.py

```diff
@@ -0,0 +1,361 @@
+                yield os.path.join(dirpath, filename)
+
+
+def find_icons(builddir, name):
+    icondir = os.path.join(builddir, 'files/share/icons/hicolor')
+    for (dirpath, dirnames, filenames) in os.walk(icondir):
+        for filename in filenames:
+            if filename.startswith(name + '.'):
+                yield os.path.join(dirpath, filename)
+
+
+def update_desktop_files(app_id, builddir):
+    for full_path in find_desktop_files(builddir):
+        cp = configparser.RawConfigParser()
+        cp.read([full_path])
+        try:
+            icon = cp.get('Desktop Entry', 'Icon')
+        except configparser.NoOptionError:
+            icon = None
+
+        # Does it have an icon?
+        if icon and not icon.startswith(app_id):
+            found_icon = False
+
+            # Rename any matching icons
+            for icon_file in find_icons(builddir, icon):
+                shutil.copy(icon_file,
+                            os.path.join(os.path.dirname(icon_file),
+                                         add_app_prefix(app_id, icon, os.path.basename(icon_file))))
+                found_icon = True
+
+            # If we renamed the icon, change the desktop file
+            if found_icon:
+                subprocess.check_call(['desktop-file-edit',
+                                       '--set-icon',
+                                       add_app_prefix(app_id, icon, icon), full_path])
+
+        # Is the desktop file not prefixed with the app id, then prefix it
+        basename = os.path.basename(full_path)
+        if not basename.startswith(app_id):
+            shutil.move(full_path,
+                        os.path.join(os.path.dirname(full_path),
+                                     add_app_prefix(app_id,
+                                                    basename[:-len('.desktop')],
+                                                    basename)))
+
+
+class FlatpakCreateOciPlugin(PrePublishPlugin):
+    key = 'flatpak_create_oci'
+    is_allowed_to_fail = False
+
+    def __init__(self, tasker, workflow):
+        """
+        :param tasker: DockerTasker instance
+        :param workflow: DockerBuildWorkflow instance
+        """
+        super(FlatpakCreateOciPlugin, self).__init__(tasker, workflow)
+
+    # Compiles a list of path mapping rules to a simple function that matches
+    # against a list of fixed patterns, see below for rule syntax
+    def _compile_target_rules(rules):
+        ROOT = "var/tmp/flatpak-build"
+
+        patterns = []
+        for source, target in rules:
+            source = re.sub("^ROOT", ROOT, source)
+            if source.endswith("/"):
+                patterns.append((re.compile(source + "(.*)"), target))
+                patterns.append((source[:-1], target))
+            else:
+                patterns.append((source, target))
+
+        def get_target_func(self, path):
+            for source, target in patterns:
+                if isinstance(source, str):
```

Comment:
This is slightly fragile because of the differences between Python 2 (`str` and `unicode`) and Python 3. Possibly more robust approach:
```
try:
    m = source.match(path)
    if m:
        return os.path.join(target, m.group(1))
except AttributeError:
    # Not a regular expression, just a filename
    if source == path:
        return target
```

Reply:
Good point to be cautious about `isinstance(x, str)`, but since the rules are hardcoded in the source code, we're just counting that `isinstance("somestring", str)`, which should be safe both python2 and python3. I'd be inclined to leave it as is, but if you want, I'll switch to compiling patterns for everything and skip the literal-matches-as-strings "optimization". Rough timings

  1.5us - isinstance(str) and string compare
  4.7us - try match, catch AttributeError, string compare
  3.5us - compile a ^foo$ regex and just use that

 (I think we'd do max. 100,000 rule / filename matches in any case, so the worst/best difference is < .3s, and not a big deal.)

Reply:
Maybe using a pattern for everything would be better then. As it stands, even with hardcoded strings, a `from __future__ import unicode_literals` statement (used in lots of other modules) would break it.

Using patterns even for fixed strings might make the code easier to follow anyway.